File: Kafka Ka Kalesh/d1.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem: Kafka Ka Kalesh
 * 
 * Producer writes 1 message every P seconds, starting at t = P.
 * Consumer tries to read 1 message every C seconds, starting at t = S.
 * Consumer can only read if there is at least one message in the buffer.
 * If a write and read happen at the same time, write happens first.
 * Given C >= P, find number of unread messages at time L.
 *
 * Approach:
 *  - Total messages produced by time L: floor(L / P)
 *  - Consumer attempt times: S, S+C, S+2C, ... <= L
 *    Number of attempts = 0 if S > L, else floor((L - S) / C) + 1
 *  - Any attempt before the first message (t < P) fails.
 *    Since C >= P, at most the very first attempt can fail.
 *  - So:
 *      if S < P and S <= L: first attempt fails, others succeed
 *      else: all attempts succeed
 *  - Lag = produced - successful_attempts
 */

void solve() {
    long long P, C, S, L;
    cin >> P >> C >> S >> L;

    // Total messages produced by time L
    long long produced = L / P;

    // Count consumer attempts in [0, L]
    long long attempts = 0;
    if (S <= L) {
        attempts = (L - S) / C + 1;
    }

    // Determine how many of these attempts actually succeed
    long long successful = attempts;

    // If the consumer starts before the first message and has at least one attempt,
    // then the first attempt fails.
    if (S < P && S <= L) {
        successful = max(0LL, attempts - 1);
    }

    // Unread messages = produced - successfully consumed
    long long lag = produced - successful;
    cout << lag << "\n";
}

int main() {
    // Fast I/O
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    while (N--) {
        solve();
    }
    return 0;
}

File: Kafka Ka Kalesh/d1.py
import sys

"""
Problem: Kafka Ka Kalesh

Producer:
  - Writes 1 message every P seconds.
  - First message at t = P, then 2P, 3P, ...

Consumer:
  - Tries to read 1 message every C seconds.
  - First attempt at t = S, then S+C, S+2C, ...
  - Can only read if buffer >= 1; otherwise skips that attempt.
  - If write and read happen at same time, write happens first.

Given C >= P, find number of unread messages at time L.

Approach:
  - produced = floor(L / P)
  - attempts = number of k >= 0 such that S + k*C <= L
             = 0 if S > L, else floor((L - S) / C) + 1
  - Because C >= P, at most the first attempt can fail.
    It fails only if S < P and S <= L.
  - successful = attempts, or attempts - 1 in that case.
  - lag = produced - successful
"""

def solve():
    data = sys.stdin.read().strip().split()
    if not data:
        return

    it = iter(data)
    N = int(next(it))
    out = []

    for _ in range(N):
        P = int(next(it))
        C = int(next(it))
        S = int(next(it))
        L = int(next(it))

        # Total messages produced by time L
        produced = L // P

        # Number of consumer attempts in [0, L]
        attempts = 0
        if S <= L:
            attempts = (L - S) // C + 1

        # Determine how many attempts actually succeed
        successful = attempts

        # If consumer starts before first message and has at least one attempt,
        # the first attempt fails.
        if S < P and S <= L:
            successful = max(0, attempts - 1)

        lag = produced - successful
        out.append(str(lag))

    sys.stdout.write("\n".join(out) + "\n")

if __name__ == "__main__":
    solve()